### PR TITLE
RFC: Rewrite the IO namespace

### PR DESCRIPTION
**Pull request by fredemmott**
This is largely based on `hh-clilib`, and using its' abstractions in HHAST (interactive repl + language server protocol implementation) and hacktest (traditional output)

- remove locking functionality: will bring it back later, just trying to
  get a clean MVP for now
- async everything
- blocking accessors
- IO hierarchy, including support for user-space implementations
- separarate read, write, and readwrite interfaces
- directly expose stdin/out/err
- FileMode enums are opaque as the values appear to be unique to PHP

# Interfaces

- Handle (root)
- ReadHandle extends Handle
- WriteHandle extends Handle
- ReadWriteHandle extends ReadHandle, WriteHandle
- FileHandle extends Handle
- FileReadHandle extends FileHandle and .... for write and readwrite
- UserspaceHandle implements Handle (expect implementations to do more)

# Classes: all private

- NativeHandle implements ReadWriteHandle
- StdioHandle extends NativeHandle
- FileHandle extends NativeHandle implements FileReadWriteHandle

### CODE REVIEW

**Comment by fredemmott** [comment]
Sockets are currently missing, but should be straightforward to add on top of `NativeFileHandle` in a similar manner to `FileHandle` and `StdioHandle`.

I also have a diff up (D9916783) adding a `pipe()` primitive to HHVM.

**Comment by BobertForever** [comment]
I really strongly dislike how these are not disposables and the replacement doesn't come with a way to do so

**Comment by fredemmott** [comment]
The problem with disposable is that it effectively bans any kind of storage/sharing of handle; this is the point, but is a pain for logging, multiple client sockets, etc - basically anything other than temporary files in a web context. This also effectively bans 'not really storage' situations such as creating an `Terminal` abstraction that includes stdout/stderr/stdin handles.

Providing a way to do so: yes, there should be one. Not sure what it will be yet.

**Comment by fredemmott** [comment]
We also can't reasonably have 'auto close on dispose' as generic behavior: this is not appropriate for handles like STDOUT. It would also be surprising to have it for some kinds but not others (with the possible exception of TemporaryFile when that's added)

File: src/filesystem/FileWriteMode.php [hunk 1]
```diff
@@ -0,0 +1,43 @@
...
+ *
+ *  This source code is licensed under the MIT license found in the
+ *  LICENSE file in the root directory of this source tree.
+ *
+ */
+
+namespace HH\Lib\Experimental\Filesystem;
+
+enum FileWriteMode : string {
+	/**
```
Comment (usox) [inline src/filesystem/FileWriteMode.php]: a codestyle hiccup

**Comment by fredemmott** [comment]
Tests will pass tomorrow; `IO\pipe()` is a very convenient way to test these, but only got added this morning.

**Comment by fredemmott** [comment]
out

**Pull request closed without merging**